File: note/algorithm/sort_search/quick_sort/quick_sort.py
def quick_sort(arr):
    if len(arr) <= 1: # 만약 처리할 정렬이 아니면
        return arr # 그냥 반환 (처리할 정렬이면 처리 후 재귀 호출 => 이 조건에 걸릴때까지)
    pivot = arr[len(arr)//2] # 피벗 인덱스
    lesser_arr, equal_arr, greater_arr = [], [], []
    for n in arr:
        if pivot > n:
            lesser_arr += [n] # 피벗보다 작은 값을 lesser_arr에 추가
        elif pivot < n:
            greater_arr += [n] # 피벗보다 큰 값을 greater_arr에 추가
        else:
            equal_arr += [n] # 피벗과 같은 값은 equal_arr에 추가
    return quick_sort(lesser_arr) + equal_arr + quick_sort(greater_arr) # equal_arr을 제외한 값들을 전부 재귀호출함

if __name__ == '__main__':
    print(quick_sort([3, 99, 6, 8, 9, 3, 5, 11, 16, 17]))

File: notes/algorithm/sort_search/quick_sort/quick_sort.c
#include <stdio.h>
#define swap(a, b) {int tmp = a; a = b; b = tmp;}

void quick_sort(int arr[], int init, int fin) {
  if (init >= fin) return;
  printf("init %d fin %d\n", init, fin);
  int pivot = (init+fin)/2, left = init, right = fin;
  while (left <= right) {
    while (arr[left] < arr[pivot]) left++;
    while (arr[pivot] < arr[right]) right--;
    if (left <= right) {
      if (left != right) {
        swap(arr[left], arr[right]);
      }
      left++; right--;
    }
  }
  quick_sort(arr, init, right);
  quick_sort(arr, left, fin);
}

int main() {
  int n = 10, arr[] = {10, 6, 3, 2, 5, 6, 9, 1, 3 ,10};
  for (int i = 0; i < n; i++) printf("%d ", arr[i]);
  printf("\n");
  quick_sort(arr, 0, 9);
  for (int i = 0; i < n; i++) printf("%d ", arr[i]);
  printf("\n");
  return 0;
}
